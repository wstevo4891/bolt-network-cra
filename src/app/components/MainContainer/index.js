import React, { Component } from 'react';

import './styles.scss'

export default class MainContainer extends Component {
  state = {
    width: null,
    slideLength: null
  }

  slideLengthIndex = {
    1400: 6,
    1100: 5,
    800: 4,
    500: 3
  }

  breakpoints = [1400, 1100, 800, 500]

  render() {
    if (this.state.slideLength === null) return null

    return (
      <div id="main-container">
        {this.props.children(this.state)}
      </div>
    )
  }

  componentDidMount() {
    if (this.state.slideLength === null) {
      this.updateSlideLength()
    }

    window.addEventListener("resize", this.updateSlideLength.bind(this))
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.updateSlideLength.bind(this))
  }

  updateSlideLength = () => {
    const width = window.innerWidth
    let num = null

    for (let point of this.breakpoints) {
      if (width >= point) {
        num = this.slideLengthIndex[point]
        break
      }
    }

    if (num == null) num = 2

    this.setState({
      width: width,
      slideLength: num
    })
  }
}
