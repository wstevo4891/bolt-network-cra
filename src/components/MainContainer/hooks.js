import { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import { API } from 'store'

export function useAPI() {
  const dispatch = useDispatch()

  return () => {
    dispatch(API.slideLength.set())
  }
}

export function useSlideLengthManager() {
  const slideLength = useSelector(API.slideLength.get)

  const setSlideLength = useAPI()

  useEffect(() => {
    setSlideLength()

    window.addEventListener('resize', setSlideLength)

    return () => window.removeEventListener('resize', setSlideLength)
  }, [setSlideLength])

  return slideLength
}
