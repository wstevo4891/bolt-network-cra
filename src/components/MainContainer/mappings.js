// import { API } from 'store'

export const mapStateToProps = (state) => {
  const { slideLength } = state

  if (slideLength === undefined) {
    return { slideLength: null }
  } else {
    return { slideLength }
  }
}

// export const mapDispatchToProps = (dispatch) => {
//   return {
//     setSlideLength: (dispatch) => dispatch(API.slideLength.set()),
//   }
// }
