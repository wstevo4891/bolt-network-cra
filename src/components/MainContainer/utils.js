import {
  BREAKPOINT_1400,
  BREAKPOINT_1100,
  BREAKPOINT_800,
  BREAKPOINT_500,
} from 'utils'

export const BREAKPOINTS = [
  BREAKPOINT_1400,
  BREAKPOINT_1100,
  BREAKPOINT_800,
  BREAKPOINT_500
]

export const DEFAULT = 2

export const SLIDE_LENGTH_MAP = {
  [BREAKPOINT_1400]: 6,
  [BREAKPOINT_1100]: 5,
  [BREAKPOINT_800]: 4,
  [BREAKPOINT_500]: 3,
}

export function currentSlideLength() {
  if (window === undefined) return null

  const width = window.innerWidth

  const slideLength = findSlideLength(width)

  return slideLength || DEFAULT
}

export function findSlideLength(width) {
  for (let point of BREAKPOINTS) {
    if (width >= point) {
      return SLIDE_LENGTH_MAP[point]
    }
  }
}
