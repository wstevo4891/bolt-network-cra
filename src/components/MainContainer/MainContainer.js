import PropTypes from 'prop-types'
import React from 'react'

import { useSlideLengthManager } from './hooks'

const MainContainer = ({ children }) => {
  const slideLength = useSlideLengthManager()

  if (slideLength === 0) return null

  return (
    <div id="main-container">
      {children}
    </div>
  )
}

MainContainer.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object,
  ]),
}

MainContainer.defaultProps = {
  children: null,
}

export default MainContainer
