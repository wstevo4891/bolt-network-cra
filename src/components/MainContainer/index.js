import { connect } from 'react-redux'

import { mapDispatchToProps, mapStateToProps } from './mappings'

import MainContainer from './MainContainer'

export default connect(mapDispatchToProps, mapStateToProps)(MainContainer)
