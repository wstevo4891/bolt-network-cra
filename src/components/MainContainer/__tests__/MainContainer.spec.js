import React from "react";
import { shallow, mount } from "enzyme";

import MainContainer from "../MainContainer";

global.innerWidth = 1450;

describe("<MainContainer />", () => {
  let props, wrapper;

  beforeEach(() => {
    props = {
      children: [],
      setSlideLength: jest.fn(),
      slideLength: null,
    };
  });

  describe("with default props", () => {
    beforeEach(() => {
      wrapper = shallow(<MainContainer {...props} />);
    });

    it("should not render", () => {
      expect(wrapper.exists("#main-container")).toBe(false);
    });

    it("should call setSlideLength", () => {
      expect(props.setSlideLength).toHaveBeenCalled();
    });
  });

  describe("when slideLength is defined", () => {
    beforeEach(() => {
      wrapper = shallow(<MainContainer {...props} slideLength={6} />);
    });

    it("should render", () => {
      expect(wrapper.exists("#main-container")).toBe(true);
    });

    it("should not call setSlideLength", () => {
      expect(props.setSlideLength).not.toHaveBeenCalled();
    });

    it("should match inline snapshot", () => {
      expect(wrapper).toMatchInlineSnapshot(`
        <div
          id="main-container"
        />
      `);
    });
  });

  describe('when window size changes', () => {
    it('should call setSlideLength', () => {
      mount(<MainContainer {...props} slideLength={6} />);
      global.innerWidth = 500
      global.dispatchEvent(new Event('resize'))

      expect(props.setSlideLength).toHaveBeenCalled();
    })
  })
});
