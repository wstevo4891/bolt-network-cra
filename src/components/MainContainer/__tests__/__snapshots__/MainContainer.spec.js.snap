// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainContainer /> when slideLength is defined should match snapshot 1`] = `
<div>
  <div
    data-testid="main-container"
    id="main-container"
  >
    <p>
      Hello World
    </p>
  </div>
</div>
`;
